dimod/quadratic/cyqm/cyqm_template.pyx.pxi

```diff
@@ -185,16 +193,38 @@ cdef class cyQM_template(cyQMBase):
             for vi in range(length):
                 self.add_linear(vi, linear[vi])
 
-    def add_quadratic(self, u, v, bias_type bias):
-        cdef Py_ssize_t ui = self.variables.index(u)
-        cdef Py_ssize_t vi = self.variables.index(v)
-
-        if ui == vi and (self.cppqm.vartype(ui) == cppVartype.SPIN
-                         or self.cppqm.vartype(ui) == cppVartype.BINARY):
-            raise ValueError(f"{u!r} cannot have an interaction with itself")
+    cdef Py_ssize_t _add_quadratic(self, index_type ui, index_type vi, bias_type bias) except -1:
+        # note: does not test that ui, vi are valid
+
+        if ui == vi:
+            if self.cppqm.vartype(ui) == cppVartype.SPIN:
+                raise ValueError(f"SPIN variables (e.g. {self.variables[ui]!r}) "
+                                 "cannot have interactions with themselves"
+                                 )
+            if self.cppqm.vartype(vi) == cppVartype.BINARY:
```

Comment:
I think ``if self.cppqm.vartype(ui) == cppVartype.BINARY:`` is clearer than swapping between ``ui`` and ``vi`` in two sequential `if`` statements like these.